equal.go

```diff
@@ -0,0 +1,157 @@
+package ipld
+
+// DeepEqual reports whether x and y are "deeply equal" as IPLD nodes.
+// This is similar to reflect.DeepEqual, but based around the Node interface.
+//
+// Two nodes must have the same kind to be deeply equal.
+// If either node has the invalid kind, the nodes are not deeply equal.
+//
+// Two nodes of scalar kinds (null, bool, int, float, string, bytes)
+// are deeply equal as per the rules of reflect.DeepEqual.
+// In other words, their Go values are compared, as returned by methods such as
+// AsString and AsInt.
+//
+// Links are an exception, since this function does not load links.
+// As such, if either node has the link kind, the nodes are not deeply equal.
+//
+// Two nodes of recursive kinds (map, list)
+// must have the same length to be deeply equal.
+// Their elements, as reported by iterators, must be deeply equal.
+// The elements are compared in the iterator's order,
+// meaning two maps sorting the same keys differently might not be equal.
+//
+// This function panics if any Node methods return a non-nil error.
+// It only calls valid methods for each node's kind,
+// so those errors should truly be rare and unexpected.
+//
+// Note that this function is mainly intended for testing purposes.
```

Comment:
I think we might skip this mention.  It's probably true, but there's nothing particularly regrettable about this function that means we need to discourage its use, as far as I can figure.

Reply:
The panics :)

Reply:
Still.

We can document that if a node breaks contract badly (e.g. fails to answer AsString after saying Kind()==String) it panics... but that requirement is sufficiently baseline that I think it's acceptable.

Reply:
Since there are lots of places where you're having to `panic()` would it be worth just returning `(error, bool)`? or is it really important to follow the `reflect.DeepEqual` pattern here? If you did that then it'd at least not have to come with caveats and could be used for legitimate runtime purposes (I could imagine up a bunch of ways this would be really useful outside of debugging).

Reply:
The thing with returning an error is that it can only happen if a Node implementation breaks the interface contract. For example, if calling AsInt on a Kind==Int node errors.

So those errors should be truly rare and unexpected, unless I'm missing something. Panic seems fine in that way.

Arguably, the interface could have been designed in a way that errors are not part of this equation, and one simply cannot call `AsString` on an integer kind node. One could do this with optional interfaces, for example, so that AsString is only implemented by nodes which support it, and it does not return an error.

Maybe ADLs would still need an error? I'm still not super clear on that. If we think DeepEqual might be used for ADLs, and it's good for them to return errors like these, then maybe we should make this return an error after all. Though, for the average use case, the error would always be nil.